controllers/apps/operations/rebuild_instance.go

```diff
@@ -197,23 +230,284 @@ func (r rebuildInstanceOpsHandler) ReconcileAction(reqCtx intctrlutil.RequestCtx
 	opsRes *OpsResource,
-	envForRestore []corev1.EnvVar,
 	progressDetail *appsv1alpha1.ProgressStatusDetail,
+	rebuildFrom appsv1alpha1.RebuildInstance,
 	instance appsv1alpha1.Instance,
-	backupName string,
 	index int) (bool, error) {
-	insHelper, err := r.prepareInstanceHelper(reqCtx, cli, opsRes, envForRestore, instance, backupName, index)
+	inPlaceHelper, err := r.prepareInplaceRebuildHelper(reqCtx, cli, opsRes, rebuildFrom.RestoreEnv,
+		instance, rebuildFrom.BackupName, index)
 	if err != nil {
 		return false, err
 	}
-	if backupName == "" {
-		return r.rebuildInstanceWithNoBackup(reqCtx, cli, opsRes, insHelper, progressDetail)
+
+	if rebuildFrom.BackupName == "" {
+		return inPlaceHelper.rebuildInstanceWithNoBackup(reqCtx, cli, opsRes, progressDetail)
+	}
+	return inPlaceHelper.rebuildInstanceWithBackup(reqCtx, cli, opsRes, progressDetail)
+}
+
+func (r rebuildInstanceOpsHandler) rebuildInstancesWithHScaling(reqCtx intctrlutil.RequestCtx,
+	cli client.Client,
+	opsRes *OpsResource,
+	rebuildInstance appsv1alpha1.RebuildInstance,
+	compStatus *appsv1alpha1.OpsRequestComponentStatus) (int, int, error) {
+	var (
+		completedCount int
+		failedCount    int
+		err            error
+	)
+	if len(compStatus.ProgressDetails) == 0 {
+		// 1. scale out the required instances
+		r.scaleOutRequiredInstances(opsRes, rebuildInstance, compStatus)
+		return 0, 0, nil
+	}
+	for i := range opsRes.Cluster.Spec.ComponentSpecs {
+		compSpec := &opsRes.Cluster.Spec.ComponentSpecs[i]
+		if compSpec.Name != rebuildInstance.ComponentName {
+			continue
+		}
+		// 2. check if the new pods are available.
+		var instancesNeedToOffline []string
+		if completedCount, failedCount, instancesNeedToOffline, err = r.checkProgressForScalingOutPods(reqCtx,
+			cli, opsRes, rebuildInstance, compSpec, compStatus); err != nil {
+			return 0, 0, err
+		}
+
+		if len(instancesNeedToOffline) > 0 {
+			// 3. offline the instances that require rebuilding when the new pod successfully scales out.
+			r.offlineSpecifiedInstances(compSpec, opsRes.Cluster.Name, instancesNeedToOffline)
+		}
+		break
+	}
+	return completedCount, failedCount, nil
+}
+
+func (r rebuildInstanceOpsHandler) scaleOutRequiredInstances(opsRes *OpsResource,
+	rebuildInstance appsv1alpha1.RebuildInstance,
+	compStatus *appsv1alpha1.OpsRequestComponentStatus) {
+	// 1. sort the instances
+	slices.SortFunc(rebuildInstance.Instances, func(a, b appsv1alpha1.Instance) bool {
+		return a.Name < b.Name
+	})
+
+	// 2. assemble the corresponding replicas and instances based on the template
+	rebuildInsWrapper := r.getRebuildInstanceWrapper(opsRes, rebuildInstance)
+
+	compName := rebuildInstance.ComponentName
+	lastCompConfiguration := opsRes.OpsRequest.Status.LastConfiguration.Components[compName]
+
+	for i := range opsRes.Cluster.Spec.ComponentSpecs {
+		compSpec := &opsRes.Cluster.Spec.ComponentSpecs[i]
+		if compSpec.Name != compName {
+			continue
+		}
+		if *lastCompConfiguration.Replicas != compSpec.Replicas {
```

Comment:
If a component's replicas are modified by another operation, such as h-scale ops, or if a user manually adjusts the replicas in the Cluster CR, could this lead to unintended consequences?

Reply:
by default, opsRequest is run by one. when the user manually modifies the cluster or forces run h-scale opsRequest, it maybe hang.

Reply:
It's better to mark the ops failed?